patroni/config_generator.py

```diff
@@ -0,0 +1,465 @@
+# (we use "username" in the Patroni config for some reason, other parameter names are the same).
+_AUTH_ALLOWED_PARAMETERS_MAPPING = {
+    'user': 'PGUSER',
+    'password': 'PGPASSWORD',
+    'sslmode': 'PGSSLMODE',
+    'sslcert': 'PGSSLCERT',
+    'sslkey': 'PGSSLKEY',
+    'sslpassword': '',
+    'sslrootcert': 'PGSSLROOTCERT',
+    'sslcrl': 'PGSSLCRL',
+    'sslcrldir': 'PGSSLCRLDIR',
+    'gssencmode': 'PGGSSENCMODE',
+    'channel_binding': 'PGCHANNELBINDING'
+}
+_NO_VALUE_MSG = '#FIXME'
+
+
+def get_address() -> Tuple[str, str]:
+    """Try to get hostname and the ip address for it returned by :func:`~socket.gethostname`.
+
+    .. note::
+        Can also return local ip.
+
+    :returns: tuple consisting of the hostname returned by :func:`~socket.gethostname`
+        and the first element in the sorted list of the addresses returned by :func:`~socket.getaddrinfo`.
+        Sorting guarantees it will prefer IPv4.
+
+    :raises:
+        :exc:`~patroni.exceptions.PatroniException`: if :exc:`OSError` occured.
+    """
+    hostname = None
+    try:
+        hostname = socket.gethostname()
+        return hostname, sorted(socket.getaddrinfo(hostname, 0, socket.AF_UNSPEC, socket.SOCK_STREAM, 0),
+                                key=lambda x: x[0])[0][4][0]
+    except OSError as err:
+        raise PatroniException(f'Failed to define ip address: {err}')
+
+
+class AbstractConfigGenerator(abc.ABC):
+    """Object representing the generated Patroni config.
+
+    :ivar output_file: full path to the output file to be used.
+    :ivar pg_major: integer representation of the major PostgreSQL version.
+    :ivar config: dictionary used for the generated configuration storage.
+    """
+
+    _HOSTNAME, _IP = get_address()
+    _TEMPLATE_CONFIG: Dict[str, Any] = {
+        'scope': _NO_VALUE_MSG,
+        'name': _HOSTNAME,
+        'postgresql': {
+            'data_dir': _NO_VALUE_MSG,
+            'connect_address': _NO_VALUE_MSG + ':5432',
+            'listen': _NO_VALUE_MSG + ':5432',
+            'bin_dir': '',
+            'authentication': {
+                'superuser': {
+                    'username': 'postgres',
+                    'password': _NO_VALUE_MSG
+                },
+                'replication': {
+                    'username': 'replicator',
+                    'password': _NO_VALUE_MSG
+                }
+            }
+        },
+        'restapi': {
+            'connect_address': _IP + ':8008',
+            'listen': _IP + ':8008'
+        }
+    }
+
+    def __init__(self, output_file: Optional[str], dsn: Optional[str] = None) -> None:
+        """Set up the output file (if passed), helper vars and the minimal config structure.
+
+        :param output_file: full path to the output file to be used.
+        :param dsn: DSN string for the local instance to get GUC values from.
+        """
+        self.output_file = output_file
+        self.pg_major = 0
+        self.dsn = dsn
+        self.parsed_dsn = {}
```

Comment:
SampleConfigGenerator doesn't use these attributes.

They should be moved to `RunningClusterConfigGenerator.__init__()`, _before_ calling `super().__init__()`.